aioesphomeapi/connection.py

```diff
@@ -545,115 +525,83 @@ async def send_message_await_response(
 
         return res[0]
 
-    async def _report_fatal_error(self, err: Exception) -> None:
-        """Report a fatal error that occurred during an operation.
+    def _handle_fatal_error(self, err: Exception) -> None:
+        """Handle a fatal error that occurred during an operation."""
+        self._connection_state = ConnectionState.CLOSED
+        for handler in self._read_exception_handlers[:]:
+            handler(err)
+        self._read_exception_handlers.clear()
+
+    def _report_fatal_error_and_cleanup_task(self, err: Exception) -> None:
+        """Handle a fatal error that occurred during an operation.
 
         This should only be called for errors that mean the connection
         can no longer be used.
 
         The connection will be closed, all exception handlers notified.
         This method does not log the error, the call site should do so.
         """
-        self._connection_state = ConnectionState.CLOSED
-        for handler in self._read_exception_handlers[:]:
-            handler(err)
-        await self._cleanup()
-
-    async def _process_loop(self) -> None:
-        to_process = self._to_process
-        while True:
-            try:
-                pkt = await to_process.get()
-            except RuntimeError:
-                break
-
-            if pkt is None:
-                # Socket closed but task isn't cancelled yet
-                break
-
-            msg_type_proto = pkt.type
-            if msg_type_proto not in MESSAGE_TYPE_TO_PROTO:
-                _LOGGER.debug(
-                    "%s: Skipping message type %s", self.log_name, msg_type_proto
-                )
-                continue
-
-            msg = MESSAGE_TYPE_TO_PROTO[msg_type_proto]()
-            try:
-                msg.ParseFromString(pkt.data)
-            except Exception as e:
-                _LOGGER.info(
-                    "%s: Invalid protobuf message: type=%s data=%s: %s",
-                    self.log_name,
-                    pkt.type,
-                    pkt.data,
-                    e,
-                    exc_info=True,
-                )
-                await self._report_fatal_error(
-                    ProtocolAPIError(f"Invalid protobuf message: {e}")
-                )
-                raise
+        self._handle_fatal_error(err)
+        asyncio.create_task(self._cleanup())
```

Comment:
`self._cleanup` doesn't do any await and could also be made non-async.

Reply:
Good idea. I'll do a follow up to address that

Reply:
Started a PR here https://github.com/esphome/aioesphomeapi/pull/355 

Its a bit more involved that I was expecting but should still finish it up today